### PR TITLE
Add Fragment Node Creation for Named Fragments

### PR DESCRIPTION
**Pull request by DimuthuMadushan**
## Purpose

Fixes: [#1922](https://github.com/ballerina-platform/ballerina-standard-library/issues/1922)

## Examples
N/A
## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: ballerina/validator_visitor.bal [hunk 1]
```diff
@@ -234,24 +226,22 @@ class ValidatorVisitor {
...
+            __Type? requiredType = getTypeFromTypeArray(<__Type[]>parentType?.possibleTypes, selection.getOnType());
             if requiredType is __Type {
                 __Field subField = createField(parentField.name, requiredType);
-                self.visitFragment(fragmentNode, subField);
+                self.visitFragment(selection, subField);
             } else {
-                string message = getFragmetCannotSpreadError(fragmentNode, selection.name, parentType);
-                self.errors.push(getErrorDetailRecord(message, <Location>selection?.spreadLocation));
+                string message = getFragmetCannotSpreadError(selection, selection.getName(), parentType);
+                self.errors.push(getErrorDetailRecord(message, <Location> selection.getSpreadLocation()));
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:244]:
```suggestion
                self.errors.push(getErrorDetailRecord(message, <Location>selection.getSpreadLocation()));
```

File: native/src/main/java/io/ballerina/stdlib/graphql/runtime/engine/ResponseGenerator.java [hunk 2]
```diff
@@ -97,10 +96,9 @@ static void getDataFromService(ExecutionContext executionContext, BObject servic
         } else {
             executionContext.setTypeName(service.getType().getName());
             for (int i = 0; i < selections.size(); i++) {
-                BMap<BString, Object> selection = (BMap<BString, Object>) selections.get(i);
-                boolean isFragment = selection.getBooleanValue(IS_FRAGMENT_FIELD);
-                BObject subNode = selection.getObjectValue(NODE_FIELD);
-                if (isFragment) {
+                BObject subNode = (BObject) selections.get(i);
+                BString typeName = StringUtils.fromString(subNode.getType().getName());
+                if (typeName.equals(FRAGMENT_NODE)) {
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/graphql/runtime/engine/ResponseGenerator.java:101]: Usually, when comparing two objects using the `equals` method, use the constant first. It will reduce the possibility of throwing an NPE.
```suggestion
                if (FRAGMENT_NODE.equals(typeName)) {
```

File: changelog.md [hunk 3]
```diff
@@ -17,6 +17,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ### Changed
 - [[#1597] Validate Max Query Depth at Runtime](https://github.com/ballerina-platform/ballerina-standard-library/issues/1597)
+- [[#1922] Create Fragment Node for Named Fragments](https://github.com/ballerina-platform/ballerina-standard-library/issues/1922)
```
Comment (ThisaruGuruge) [inline changelog.md:20]: How about rewording like this? 
```suggestion
- [[#1922] Unify Parsing Named Fragments and Inline Fragments](https://github.com/ballerina-platform/ballerina-standard-library/issues/1922)
```

File: ballerina/duplicate_field_remover.bal [hunk 4]
```diff
@@ -78,107 +70,46 @@ class DuplicateFieldRemover {
...
-        while (i < selections.length()) {
+        while i < selections.length() {
             parser:Selection selection = selections[i];
-            if(selection.isFragment) {
-                parser:FragmentNode fragmentNode = self.documentNode.getFragments().get(selection.name);
-                if (visitedFragmentOnTypes.hasKey(fragmentNode.getOnType())) {
-                    if (fragmentNode.isInlineFragment()) {
+            if selection is parser:FragmentNode {
+                if visitedFragmentOnTypes.hasKey(selection.getOnType()) {
+                    if selection.isInlineFragment() {
```
[the remark this answers is not in the document]
Comment (DimuthuMadushan) [reply ballerina/duplicate_field_remover.bal]: change to process both of the fragments similarly.

File: ballerina/fragment_visitor.bal [hunk 5]
```diff
@@ -87,15 +84,26 @@ class FragmentVisitor {
...
     }
 
-    isolated function getFragment(parser:Selection selection) returns parser:FragmentNode? {
+    isolated function appendNamedFragmentFields(parser:FragmentNode selection) {
         parser:DocumentNode documentNode = <parser:DocumentNode>self.documentNode;
-        parser:FragmentNode? fragmentNode = documentNode.getFragment(selection.name);
-        if (fragmentNode is ()) {
-            string message = string`Unknown fragment "${selection.name}".`;
-            ErrorDetail errorDetail = getErrorDetailRecord(message, selection.location);
+        parser:FragmentNode? fragmentNode = documentNode.getFragment(selection.getName());
```
[the remark this answers is not in the document]
Comment (DimuthuMadushan) [reply ballerina/fragment_visitor.bal]: Since the `selection` is a fragment node, this is correct. isn't it?

File: ballerina/duplicate_field_remover.bal [hunk 6]
```diff
@@ -78,107 +70,41 @@ class DuplicateFieldRemover {
...
-        }
-        foreach string fragments in duplicate.getFragments() {
-            original.addFragment(fragments);
-        }
-        foreach parser:Selection selections in duplicate.getSelections() {
-            original.addSelection(selections);
+        foreach parser:Selection selection in duplicate.getSelections() {
+            original.addSelection(selection);
         }
     }
```
Comment (ThisaruGuruge) [inline ballerina/duplicate_field_remover.bal:109]: Can't we combine these two functions?

File: ballerina/fragment_visitor.bal [hunk 7]
```diff
@@ -87,15 +84,26 @@ class FragmentVisitor {
...
+        } else {
+            if !fragmentNode.isInlineFragment() && fragmentNode.getSelections().length() == 0 {
+                self.appendFields(actualFragmentNode, fragmentNode);
+            }
+        }
+    }
+
+    isolated function appendFields(parser:FragmentNode actualFragmentNode, parser:FragmentNode fragmentNode) {
+        fragmentNode.setOnType(actualFragmentNode.getOnType());
+        fragmentNode.setLocation(actualFragmentNode.getLocation());
```
Comment (ThisaruGuruge) [inline ballerina/fragment_visitor.bal:103]: Do we need this location? Isn't `spreadLocation` enough?

File: ballerina/fragment_visitor.bal [hunk 8]
```diff
@@ -87,15 +84,26 @@ class FragmentVisitor {
...
-            ErrorDetail errorDetail = getErrorDetailRecord(message, selection.location);
+    isolated function appendNamedFragmentFields(parser:FragmentNode fragmentNode) {
+        parser:DocumentNode documentNode = self.documentNode;
+        parser:FragmentNode? actualFragmentNode = documentNode.getFragment(fragmentNode.getName());
+        if actualFragmentNode is () {
+            string message = string`Unknown fragment "${fragmentNode.getName()}".`;
+            ErrorDetail errorDetail = getErrorDetailRecord(message, fragmentNode.getLocation());
             self.errors.push(errorDetail);
+        } else {
+            if !fragmentNode.isInlineFragment() && fragmentNode.getSelections().length() == 0 {
```
Comment (ThisaruGuruge) [inline ballerina/fragment_visitor.bal:95]: What's the reason for using this condition `fragmentNode.getSelections().length() == 0` ?

File: ballerina/fragment_visitor.bal (hunk 5 above)
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply ballerina/fragment_visitor.bal]: Fragments too can have aliases right?

File: ballerina/fragment_visitor.bal (hunk 8 above)
Comment (DimuthuMadushan) [reply ballerina/fragment_visitor.bal:95]:
```graphql
query A($profileId:Int!) {
    ...data 
}

query B($profileId:Int!) {
    ...data
}

fragment data on Query {
    profile(id:$profileId) {
        ...people
    }
}

fragment people on Person {
    address {
        city
    }
}
```
when visiting a document like this, `selection[]` of  `...people` node in fragment `data` is also getting filled while procesing the `...data` node `selection[]` in query `A`. here,  `fragmentNode.getSelections().length() == 0` condition is used to avoid re-filling the `selection[]` in `...people` node when processing the `...data` node in query `B`.

**Review by ThisaruGuruge — APPROVED** [review]

**Review by ThisaruGuruge — APPROVED** [review]
